auto_round/utils/device.py

```diff
@@ -628,8 +598,179 @@ def set_non_auto_device_map(
+            - layer_memory_dict (dict): A dictionary mapping layer names to their memory consumption (in GB).
+              Format: {layer_name: {"param_memory": float, "output_memory": float}}
+            - input_output_memory (float): The memory consumption (in GB) for input and output
+                tensors of the block.
+            - additional_memory (float): Additional memory overhead (in GB) for operations like attention.
+    """
+    # Calculate all block parameters memory and build layer-wise memory dict
+    from auto_round.utils.model import get_layer_features
+
+    layer_memory_dict = {}
+    total_param_mem = 0
+
+    # Calculate batch_size and sequence_length from input_ids for output memory estimation
+    seq_len = input_ids[0].shape[1] if input_ids and len(input_ids[0].shape) >= 2 else 1
+    element_size = input_ids[0].element_size() if input_ids else 2  # Default to 2 bytes (fp16/bf16)
+
+    for name, module in block.named_modules():
+        if check_to_quantized(module):
+            enable_act_quant = module.act_bits <= 8
+            layer_name = name
+            param_size = module.weight.nbytes
+            param_memory_gb = param_size / 1024**3
+            param_memory_gb *= 2  # considering the v tensor for weight rounding
+
+            # Estimate output memory based on input_features and out_features
+            in_features, out_features = get_layer_features(module)
+            if in_features is not None and out_features is not None:
+                # Output tensor size: batch_size * seq_len * out_features * element_size
+                output_size = pick_samples * seq_len * out_features * element_size
+                output_memory_gb = output_size / 1024**3
+
+                # If enable_act_quant, add input tensor memory to param_memory
+                if enable_act_quant:
+                    input_size = pick_samples * seq_len * in_features * element_size
+                    input_memory_gb = input_size / 1024**3
+                    param_memory_gb += input_memory_gb
+            else:
+                output_memory_gb = 0.0
+
+            # memory * 2, because it contains grad tensor.
+            layer_memory_dict[layer_name] = {"param_memory": param_memory_gb * 2, "output_memory": output_memory_gb * 2}
+
+    # Assuming bfloat16 or float32, input and output
+    block_input_output_memory = 2 * sum(tensor.nbytes for tensor in input_ids) / 1024**3
+
+    # Roughly estimate additional memory for attention and other operations
+    additional_activation_memory = sum(info["output_memory"] for info in layer_memory_dict.values())
+    # 1GB considers norm weight, sdpa, reference_output, etc.
+    additional_memory = additional_activation_memory + 1  # GB
+    if torch.xpu.is_available():
+        # https://github.com/intel/torch-xpu-ops/issues/2232
+        # TODO: XPU takes more memory than expected. for llama 8B, it's about 12 GB
+        xpu_additional_memory = 12  # GB
```

Comment:
Could you explain more about where the 12GB comes from? Is t related to the model? 
Would that be okay for 100B+ models?

Reply:
It's not okay, It's for 8b on XPU, and we should expect this issue is fixed in WW47